### PR TITLE
Finding bounding boxes of substrings of extracted text.

### PR DESCRIPTION
**Pull request by peterwilliams97**
This PR combines the TextComponent idea from https://github.com/unidoc/unidoc/compare/v3...gunnsth:giggiu16-text-extraction-origtext with a mechanism for mapping substrings in extracted text positions described in the comments below.

Addresses parts of #35.

// TextMark maps extracted text to the location of the text on the PDF page and other
// properties of the rendered text such the font.
// `Offset` is the offset of the start of the textMark.text in extracted text.
// `BBox` is the bounding box of the textMark.
// `Text` is the extracted text
// `Meta` is set true for characters that don't appear in the input PDF.
// You can find the location of substrings in the extracted text as follows:
//   Use ToTextLocation() to return the extracted text as a []TextMark sorted by Offset.
//   substring := extracted[start:end+1]
//   binary search the []TextMark for `start` and `end`.
//   The bounding box of `substring` on the PDF page is the union of the TextMark.BBox's with
//     start <= TextMark.Offset < end
//
// getBBox() in test_text.go shows how to compute bounding boxes of substrings of extracted text.
// The following code extracts the text on PDF page `page` into `text` then finds the bounding box
// `bbox` of substring `term` in `text`. (indexRunes() works like strings.Index except that it
// returns number of runes rather than numberof bytes)
//
//     ex, _ := New(page)
//     // handle errors
//     pageText, _, _, err := ex.ExtractPageText()
//     // handle errors
//     text := pageText.Text()
//     textMarks := pageText.Marks()
//
//     start := indexRunes(text, term)
//     end := start + len([]rune(term))
//     spanMarks, err := textMarks.RangeOffset(start, end)
//     // handle errors
//     bbox, ok := spanMarks.BBox()
//     // handle errror

---
This change is [Reviewable](https://reviewable.io/reviews/unidoc/unipdf/109)

### CODE REVIEW

**Comment by peterwilliams97** [comment]
___
*[extractor/text.go, line 782 at r4](https://reviewable.io/reviews/unidoc/unipdf/109#-Lj1o9d1CsDz2IdapTzJ:-LjPEsyL34CkjQGseerD:b-qs3ojp) ([raw file](https://github.com/unidoc/unipdf/blob/a9e8aad1a661a228adf26ad2e1ca958210955ba8/extractor/text.go#L782)):*
<details><summary><i>Previously, gunnsth (Gunnsteinn Hall) wrote…</i></summary><blockquote>

It would be good to clarify the difference between a textMark, TextMark and TextComponent.
To me it appears that roughly there is one textMark for each operand roughly?  In many cases
thus one letter could be represented by a textMark.  It's common to use TJ to draw single letters with custom spacing between each letter for instance.
Would that lead to multiple textMarks, whereas when converted to TextComponent it would try to group more of those together?

I guess overall, it would be good to define exactly what a "textMark" is in terms of how it works.
</blockquote></details>

I can do that in the code but can we discuss it here first?
 textMark, TextMark and TextComponent all represent the same thing. A single rendered glyph in device coordinates. 
 I think we need a private view, textMark, to decouple our internal development from users' development.
Currently I only see a need for one public view. I use TextComponent because it has everything I need.
I don't see a need for TextMark
There may be value in extending TextComponent or adding more public views in the future.

**Comment by peterwilliams97** [comment]
___
*[extractor/text.go, line 900 at r4](https://reviewable.io/reviews/unidoc/unipdf/109#-Lj1p9OqEo8UGJONlw_z:-LjPGT9Q9-IpYJwnrOvi:b-bqk2og) ([raw file](https://github.com/unidoc/unipdf/blob/a9e8aad1a661a228adf26ad2e1ca958210955ba8/extractor/text.go#L900)):*
<details><summary><i>Previously, gunnsth (Gunnsteinn Hall) wrote…</i></summary><blockquote>

It would be better to assume the outside user has no idea what "textMark" is and explain what TextMark is and how they can use it.
</blockquote></details>

Will do.

**Comment by peterwilliams97** [comment]
___
*[extractor/text.go, line 908 at r4](https://reviewable.io/reviews/unidoc/unipdf/109#-Lj1pLBICVd5KFOSrS10:-LjPGY_d9IHF-E1XBTyS:b-cstnkg) ([raw file](https://github.com/unidoc/unipdf/blob/a9e8aad1a661a228adf26ad2e1ca958210955ba8/extractor/text.go#L908)):*
<details><summary><i>Previously, gunnsth (Gunnsteinn Hall) wrote…</i></summary><blockquote>

Probably not necessary if TextComponent provides equivalent functionality and more?
</blockquote></details>

I agree

**Comment by peterwilliams97** [comment]
___
*[extractor/text.go, line 960 at r4](https://reviewable.io/reviews/unidoc/unipdf/109#-Lj1q5T6Cqd_Hh5PMxdl:-LjPGohp6P5SDh0PKsW-:b-vz9yj6) ([raw file](https://github.com/unidoc/unipdf/blob/a9e8aad1a661a228adf26ad2e1ca958210955ba8/extractor/text.go#L960)):*
<details><summary><i>Previously, gunnsth (Gunnsteinn Hall) wrote…</i></summary><blockquote>

Can we describe this in a way that does not require knowing the internals?  Like how it can be used for instance.  Fine to mention (textMark.text) in parenthesis but in general should not require knowing the internals.

What about simply exporting textMark rather than multiple slightly higher level views on it?
</blockquote></details>

I was thinking we should wait until we have experience of a few more applications of our text extraction before we finalise our internal code.
e.g.We have yet to implement text groupings such as word, paragraphs, columns and tables.
I don't know how much of a problem this is. All rendered PDF text is composed of rendered glyphs and TextComponent is a reasonable way to represent a rendered glyph. However there are internals such as textMark.trm that I don't expect a user to ever need to know about. 
These are library design questions, so you should decide what to do.

**Comment by peterwilliams97** [comment]
___
*[extractor/text.go, line 980 at r4](https://reviewable.io/reviews/unidoc/unipdf/109#-Lj1p_TZ7LhgUDxnVDF7:-LjPK-dE5lE4qjq2UYPf:bpiq76v) ([raw file](https://github.com/unidoc/unipdf/blob/a9e8aad1a661a228adf26ad2e1ca958210955ba8/extractor/text.go#L980)):*
<details><summary><i>Previously, gunnsth (Gunnsteinn Hall) wrote…</i></summary><blockquote>

Looking at this, seems exactly the same as the TextMark except slightly different what is exported and this brings a bit more to the table it seems.
So I do think it would make sense to reduce that to a single type.  However, we can leave TextMark as it is already published and mark it for deprecation in a future major release. 
</blockquote></details>

Sounds like a good approach. 
Could we add the  extra TextComponent fields to TextMark. That shouldn't break anything and I think TextMark is a better name than a TextComponent.

**Comment by peterwilliams97** [comment]
___
*[extractor/text.go, line 1006 at r4](https://reviewable.io/reviews/unidoc/unipdf/109#-Lj1qrgk8zpmazhw6Iy6:-LjPK_bT-Aoj3KkUDqrH:b-ex0edm) ([raw file](https://github.com/unidoc/unipdf/blob/a9e8aad1a661a228adf26ad2e1ca958210955ba8/extractor/text.go#L1006)):*
<details><summary><i>Previously, gunnsth (Gunnsteinn Hall) wrote…</i></summary><blockquote>

Bit confusing when to use this or the above TextComponents() function
</blockquote></details>

TextByComponents() is the only function I use to access the []TextComponent view of a page. It allows me to map substrings of the extracted text to a rendered PDF page.  What name do you suggest?

**Comment by adrg** [comment]
Hi @peterwilliams97 and @gunnsth 

Here is my proposal for the text extraction interface. Please let me know your thoughts on it.

## TextMark

Maybe `textMark` should be exported so it would become `TextMark`. All the fields would stay private and we could export them through methods so that we could also perform calculations on them before returning the values, if necessary.
```go
// textMark -> TextMark (replacing the original TextMark struct):
type TextMark struct {
	// unexported fields
}
```

The two methods below would cover the original exported `TextMark` struct which only had two public fields: `BBox` and `Text`.

```go
func (t TextMark) BBox() model.PdfRectangle {
	return t.bbox
}

func (t TextMark) Text() string {
	return t.text
}
```

We could export a couple more fields through methods, as they could be useful.

```go
func (t TextMark) Font() *model.PdfFont {
	return t.font
}

func (t TextMark) FontSize() float64 {
	return t.fontSize
}
```

`Marks()` would simply return pt.marks.

```go
func (pt PageText) Marks() []TextMark {
        // Optionally, we could return the sorted text marks.
        // For efficiency, sorting should only be performed once.
	fontHeight := pt.height()
	tol := minFloat(fontHeight*0.2, 5.0)
	pt.sortPosition(tol)

	return pt.marks
}
```

## TextLine

`textLine` is defined like below. The `counts` field does not seem to be vital so I will leave it out for now.

```go
type textLine struct {
	x      float64              // x position of line.
	y      float64              // y position of line.
	h      float64              // height of line text.
	dxList []float64            // x distance between successive words in line.
	words  []string             // words in the line.
	bboxes []model.PdfRectangle // bboxes[i] = bounding box of words[i].
	fonts  []*model.PdfFont     // fonts[i] = the font words[i] was rendered with.
	counts []int64              // textMark count (to help with reading debug logs).
}
```

I believe we should introduce a new struct here for words, as the `PageText.toLines()` method (which is also used by `TextByComponents()`) does seem to try to group individual text marks into words.

```go
type Word struct {
	Text     string
	BBox     []model.PdfRectangle
	Font     *model.PdfFont
	FontSize float64
}
```

We could also export `textLine`, leaving the fields private.

```go
// textLine -> TextLine and also adding the Word struct:
type TextLine struct {
	x      float64   // x position of line.
	y      float64   // y position of line.
	h      float64   // height of line text.
	dxList []float64 // x distance between successive words in line.

	words []Word // the collection of words
}

func (tl TextLine) Pos() (x, y float64) {
	return tl.x, tl.y
}

func (tl TextLine) Height() float64 {
	return tl.h
}

func (tl TextLine) Words() []Word {
	return tl.words
}
```

This way we could also export a method called `Lines()` which would return all the lines of the page. The code would be something like this:

```go
func (pt PageText) Lines() []TextLine {
	fontHeight := pt.height()
	tol := minFloat(fontHeight*0.2, 5.0)
	pt.sortPosition(tol)
	return pt.toLines(tol)
}
```

`TextByComponents()` would become `Words()` but I don't see why it should also return the text. 
So maybe we can drop that as there are other ways to retrieve the page text.
`Words()` would return all words from all lines, sorted left to right, top to bottom.

```go
// TextByComponents -> Words
func (pt PageText) Words() []Word {
	var words []Word

	// Use the previously proposed method to get the sorted lines.
	lines := pt.Lines()
	for i, l := range lines {
		words = append(words, l.words...)
	}

	return words
}
```

In the original `TextComponent` struct, the `Offset` field seems to be initialized only when getting the sorted slice of text components, by calling the `PageText.toLines()` method, which seems to group text marks into words. The `Offset` field seems to be incremented by the number of runes in the word and the length of the word and line separators.
Above, we return a slice of sorted words from the `Words()` method. So is there any reason for not combining the bounding boxes of the sorted words (by their index in the slice) in order to get the bounding box of a substring?
If I'm not missing something, I think this method should not be defined at all.

```go
func GetBBox(locations []TextComponent, offset int) (model.PdfRectangle, bool) {
}
```

## PageText interface

In short, the interface of the PageText object would become:

```go
Marks() []TextMark
Lines() []TextLine
Words() []Word
ToText() string
```

As a side note, maybe the `ToText()` method should just be called `Text()`.

If you believe exporting text marks does not have any advantage, we can make it private by removing the original `TextMark` struct completely. However, we could add documentation to it and label it as unprocessed text and export it for advanced library users.

**Comment by peterwilliams97** [comment]
adrg  

That looks good. Almost everything you said is correct and I agree with most of your proposal so I am going to respond to what to the few problems I see with it.

A word can have more than one font. I don't know if users of our code will ever care but it can happen
```
type Word struct {
	Text     string
	BBox     []model.PdfRectangle
	Font     *model.PdfFont
	FontSize float64
}

For consistency, we could have a `func (tl TextLine) BBox() model.PdfRectangle`

``
The current code doesn't deal with columns of text or tables. We need an interface which will allow us to add column and table support in the future. 

 In this interface, 
```
Marks() []TextMark
Words() []Word
Lines() []TextLine
Text() string
```
 `Marks()` and `Words()` are the two lowest level objects and their meaning won't change in columns and tables. 
`Text()` gives the page text. The contents of the teext would be re-ordered with columns and tables, but it would still be a string. The function signature would not change. 
Adding extra member functions such as `Columns()` or `Cells()` in the future won't break any of the interface, but 
1) we should think about what a `Line` will mean within a `column` or a `cell`.  
2) what extra internal fields we will need to implement  `Columns()` and `Cells()`
```

**Comment by peterwilliams97** [comment]
Changing `TextByComponents ()` -> `Words()` has an effect.  A lot of text processing code takes a string (not a list of words) as input and returns `(start, end)` offsets  of matched queries in the text. Currently we construct `Line := Join(Words, " ")` and Text := Join(Lines, "\n").  `TextByComponents ()` performs these Joins in a way that allows us to look up the TextMarks that `(start, end)`  span (this is the purpose of the `Offset` field.
A user can always do this Join in his / her own code but his/her text may differ from our `Text()`
One way to achieve this consistency in the new interface would be to compute Marks(), Words(), Lines() and Text() once per page and use these member functions to access the computed values. This would be when the `Offset` field is computed

**Comment by adrg** [comment]
@peterwilliams97 

Making `TextByComponents()` private  (and renaming it)  does seem like a good solution. I believe it should be called at the end of the private `extractPageText()` method, before returning the page text.

I was thinking that maybe `TextComponent` should become `TextMark` and we could leave `textMark` for internal usage. I think this makes sense because it would not break the existing functionality, only add to it. So `Marks()` would return a collection of `TextComponent` objects, pre-computed by the `computeViews()` method at extraction time.

If the concept of line is not clearly defined at the moment, we could leave it private for now.

For measuring substrings, I like second approach:

```go
type TextBase interface {
    BBox(elements []TextBase) model.PdfRectangle
}
```
The disadvantage here is that user could combine entities (lines, marks, words) into a `TextBase` slice which would produce invalid results. 

Another one would be to define the `BBox` on a custom type represented by a slice of `TextMark` objects:

```go
type TextMarks []TextMark
func (t TextMarks) GetBBox() model.PdfRectangle {}
```

`Marks` could return this type in this case:

```go
Marks() TextMarks
```

Usage:

```go
marks := pageText.Marks()
marks.BBox()
marks[0:5].BBox()
```

In this case the interface would be:

```go
Marks() TextMarks
Text() string
```

`Words()` do not seem terribly important, but it can be implemented in a similar manner to `TextMark` by defining `TextWord` and `type TextWords []TextWord` which would have the `BBox()` method implemented. You are right about words potentially having multiple fonts. Not sure how that could be handled.

```go
Marks() TextMarks
Words() TextWords
Text() string
```

**Comment by peterwilliams97** [comment]
Looks good, except I haven't see this idiom of member functions on slices `marks[0:5].BBox()` in Go code before. Is it a common idiom?

**Comment by gunnsth** [comment]
@peterwilliams97 @adrg 
`PdfObjectArray` used to be a `[]PdfObject` and `ContentStreamOperations` is still a `[]*ContentStreamOperation` slice.

There are also uses of this in the Go libraries, for example https://golang.org/pkg/go/scanner/#ErrorList

I don't like this kind of type since its not immediately obvious that you can do `len(marks)` on it and access members by index `marks[5]`.  Also from experience I find that often times it becomes necessary to add functionality to it which usually requires storing more data, in which case it's a better idea to embed the slice inside a struct type.

For example:
```
func categorizeMarks(marks *TextMarks) {
  if marks == nil {
    return
  }
  for _, mark := range *marks {
  // ...
  }
  // vs. 
  for _, mark := range marks.Slice() {
  }

  // Get bounding box of first 5 marks.
  bbox := (*marks)[0:5].GetBBox()

  // vs.
  bbox := marks.SubMarks(0,5).GetBBox()
  // Those functions can do nil ptr, range checks etc.
}
```

**Comment by peterwilliams97** [comment]
I see the pattern
```
func (array *PdfObjectArray) Elements() []PdfObject {
func (array *PdfObjectArray) Len() int {
func (array *PdfObjectArray) Get(i int) PdfObject {
func (array *PdfObjectArray) Set(i int, obj PdfObject) error {
```
I will try to follow this.

**Comment by peterwilliams97** [comment]
Please take a look at the text extraction code. It works the way I use it at PaperCut. 
* `TextMarkArray` is read-only
* `TextMarkArray` is accessed by `Offset` now. We can add access by `X`, `Y` and other public TextMark fields in the future. It is not directly accessible by element order.

Let me know if you disagree with any of this.

Setting` markupPDFs=true` in text_test.go shows PDFs with blue rectangles marking the bounding boxes that are returned by `TextMarkArray.BBox()`. Please test with ` markupPDFs=true` and let me know if you disagree with the box placements.

**Comment by peterwilliams97** [comment]
___
*[extractor/text.go, line 1064 at r8](https://reviewable.io/reviews/unidoc/unipdf/109#-LjmvrMn18Gm9VKG8pF2:-LjngYUn2DxpFzwUX1oI:b-a9t2ru) ([raw file](https://github.com/unidoc/unipdf/blob/4cb6f6dfa812ae61704ca6b96ea05251152d0a32/extractor/text.go#L1064)):*
<details><summary><i>Previously, gunnsth (Gunnsteinn Hall) wrote…</i></summary><blockquote>

TextMark represents extracted text on a page with information regarding both textual content, formatting (font and size) and positioning.

Move the documentation for fields into the struct above each field member.
// Offset represents the offset of the text in which order it is defined internally in the content stream.  Not necessarily the
// same as it appears on the page.
Offset int

Is that correct by the way?

// BBox represents the bounding box of the text
// Text is the extracted unicode text.
// Original is the original decoded text.
// Meta is true for characters that don't appear in the input PDF.

Is Meta used for searchable PDFS (OCR) ?  By "don't appear" does it mean that they are not visible?  That might be better wording.
etc.

Add separator above the example
//
// Example: To get the location of substrings in extracted text:
// ...
Is this example up to date? Would be better to provide an Example in a test file?  Those look nice in godoc
</blockquote></details>


 Offset is the offset of the start of TextMark.Text in the extracted text. If you do this
```
	   text, textMarks := pageText.Text(), pageText.Marks()
	   marks := textMarks.Elements()
```
then `marks[i].Offset` is the offset of `marks[i].Text` in `text`.
`Meta` is for spaces and line breaks that we insert in the extracted text. The spaces and line breaks usually don't exist in the PDF (though they are allowed to). We insert spaces (line breaks) when we see characters that are over a threshold horizontal (vertical) distance apart.
I should do the Example test file. I will look up how to do that. We can add static text and regex search to the CLI tool too

**Comment by peterwilliams97** [comment]
___
*[extractor/text.go, line 1093 at r8](https://reviewable.io/reviews/unidoc/unipdf/109#-Ljmz2n75TkKcLtZfRx6:-LjnzOKD50kBez6T0Dan:b-896fix) ([raw file](https://github.com/unidoc/unipdf/blob/4cb6f6dfa812ae61704ca6b96ea05251152d0a32/extractor/text.go#L1093)):*
<details><summary><i>Previously, gunnsth (Gunnsteinn Hall) wrote…</i></summary><blockquote>

Receiver names should be consistent, below there are:
TextMark - t
PageText - pt
textLine - l

I suggest tm, pt, tl for those receivers.

</blockquote></details>

Done.

**Comment by peterwilliams97** [comment]
___
*[extractor/text.go, line 1110 at r8](https://reviewable.io/reviews/unidoc/unipdf/109#-LjmzC1b1pew8y-4tn7D:-Ljo0hR-9-6LZ2WgbuoO:b-896fix) ([raw file](https://github.com/unidoc/unipdf/blob/4cb6f6dfa812ae61704ca6b96ea05251152d0a32/extractor/text.go#L1110)):*
<details><summary><i>Previously, gunnsth (Gunnsteinn Hall) wrote…</i></summary><blockquote>

`pt.viewText` probably better than `pt`.viewText as its kind of an expression.
// computeViews processes the page TextMarks sorting by position and populates `pt.viewText` and `pt.viewMarks` which represent the text and marks in the order which it is read on the page.
might be a better explanation
</blockquote></details>

Done.

**Comment by peterwilliams97** [comment]
___
*[extractor/text.go, line 1233 at r8](https://reviewable.io/reviews/unidoc/unipdf/109#-Ljn1qGq83-MLa4K7xB1:-Ljo0tc1APJamO5a6SCc:b-896fix) ([raw file](https://github.com/unidoc/unipdf/blob/4cb6f6dfa812ae61704ca6b96ea05251152d0a32/extractor/text.go#L1233)):*
<details><summary><i>Previously, gunnsth (Gunnsteinn Hall) wrote…</i></summary><blockquote>

Comment from @adrg (#75):
> `concatention` should be `concatenation`
</blockquote></details>

Done.

**Comment by peterwilliams97** [comment]
___
*[extractor/text.go, line 1386 at r8](https://reviewable.io/reviews/unidoc/unipdf/109#-Ljn0Qc3CQB3pG7d2209:-Ljo2ZQl6sahhnVvZE_5:b-qeqnxg) ([raw file](https://github.com/unidoc/unipdf/blob/4cb6f6dfa812ae61704ca6b96ea05251152d0a32/extractor/text.go#L1386)):*
> ```
> w0 := line.words[0]
> ```
line.marks has one more element than line.dxList which we check. However the line.marks costs nothing and makes the code clearer so I will add it.

**Comment by peterwilliams97** [comment]
___
*[extractor/text.go, line 1391 at r8](https://reviewable.io/reviews/unidoc/unipdf/109#-Ljn28Bx2GfoIKp4UJ_B:-Ljo3Fdq8D922pUnbn6h:b4ojfdv) ([raw file](https://github.com/unidoc/unipdf/blob/4cb6f6dfa812ae61704ca6b96ea05251152d0a32/extractor/text.go#L1391)):*
<details><summary><i>Previously, gunnsth (Gunnsteinn Hall) wrote…</i></summary><blockquote>

Comment from @adrg (#75):
> This could potentially cause a panic if the loop gets to the last element in `dxList` as `i+1` would be out of range.
</blockquote></details>

`dxList` has one less element than there are text marks. It is the gaps between text marks. `dxList[i]` is the gap between `marks[i]` and `marks[i+1]` where mark is used for `bboxes`, `fonts` and `counts`.

**Comment by peterwilliams97** [comment]
___
*[extractor/text_test.go, line 140 at r7](https://reviewable.io/reviews/unidoc/unipdf/109#-Lja1hVS9_Rbkv2x0pnu:-Ljo4BjCARbo_sr5DjEH:b-61ig0s) ([raw file](https://github.com/unidoc/unipdf/blob/93acb08420e43cb269340e8a3fe75e00d1d8e8d5/extractor/text_test.go#L140)):*
<details><summary><i>Previously, gunnsth (Gunnsteinn Hall) wrote…</i></summary><blockquote>

TODO finished?  Seems like there is a flag for lazy that can be set to true.  What is the reason for having the flag?  Isn't it usually better to do extraction in lazy mode?  If there is a problem with lazy loading should be reported as an issue.
</blockquote></details>

I will test lazy loading again and report an issue or leave it enabled in the test code.

**Comment by peterwilliams97** [comment]
___
*[internal/transform/matrix.go, line 131 at r8](https://reviewable.io/reviews/unidoc/unipdf/109#-Ljn1RXs1FBSfImZLEnH:-LjoD7_KE59fae486kpk:b-tgmw1b) ([raw file](https://github.com/unidoc/unipdf/blob/4cb6f6dfa812ae61704ca6b96ea05251152d0a32/internal/transform/matrix.go#L131)):*
<details><summary><i>Previously, gunnsth (Gunnsteinn Hall) wrote…</i></summary><blockquote>

There was a comment on this from @adrg (#75):
> Maybe add some documentation to this method. Also, seems a bit odd that the function does not return anything, like a bool to specify if the matrix is ok. In its current form, the only purpose of the method is adding a line in the log.
</blockquote></details>

I changed it. 
This function was for finding bugs in my code during development.  It is probably useful in the future so I left the changed version of it in the code. Should I delete it?

**Comment by peterwilliams97** [comment]
___
*[extractor/text_test.go, line 140 at r7](https://reviewable.io/reviews/unidoc/unipdf/109#-Lja1hVS9_Rbkv2x0pnu:-LjoGbEN7mDLoEHBtKT-:bxpk5t3) ([raw file](https://github.com/unidoc/unipdf/blob/93acb08420e43cb269340e8a3fe75e00d1d8e8d5/extractor/text_test.go#L140)):*
<details><summary><i>Previously, peterwilliams97 (Peter Williams) wrote…</i></summary><blockquote>

I will test lazy loading again and report an issue or leave it enabled in the test code.
</blockquote></details>

Found the problem. I wasn't holding the file handle the PdfReader was based on openn

File: extractor/text.go [hunk 1]
```diff
@@ -848,38 +912,254 @@ func (pt PageText) length() int {
...
+// String returns a string describing `ma`.
+func (ma TextMarkArray) String() string {
+	n := len(ma.marks)
+	if n == 0 {
+		return "EMPTY"
+	}
+	m0 := ma.marks[0]
+	m1 := ma.marks[n-1]
+	return fmt.Sprintf("{TEXTMARKARRAY: %d elements\n\tfirst=%s\n\t last=%s}", n, m0, m1)
+
```
[the remark this answers is not in the document]
Comment (peterwilliams97) [reply extractor/text.go]: Done

File: extractor/text.go [hunk 2]
```diff
@@ -848,38 +912,254 @@ func (pt PageText) length() int {
...
+// Len returns the number of TextMarks in `ma`.
+func (ma *TextMarkArray) Len() int {
+	if ma == nil {
+		return 0
+	}
+	return len(ma.marks)
+}
+
+// RangeOffset returns the TextMarks in `ma` that have `start` <= TextMark.Offet < `end`.
```
[the remark this answers is not in the document]
Comment (peterwilliams97) [reply extractor/text.go]: Done

File: extractor/text.go [hunk 3]
```diff
@@ -848,38 +912,254 @@ func (pt PageText) length() int {
...
+func (ma *TextMarkArray) RangeOffset(start, end int) (*TextMarkArray, error) {
+	if ma == nil {
+		return nil, errors.New("ma==nil")
+	}
+	if end < start {
+		return nil, fmt.Errorf("end < start. RangeOffset not defined. start=%d end=%d ", start, end)
+	}
+	if len(ma.marks) == 0 {
```
[the remark this answers is not in the document]
Comment (peterwilliams97) [reply extractor/text.go]: I have left this. This function is likely to be called fewer times than there are characters in PDF files, not a large number. len() is efficient

File: extractor/text.go [hunk 4]
```diff
@@ -848,38 +912,254 @@ func (pt PageText) length() int {
...
+		return nil, err
+	}
+	if iEnd <= iStart {
+		// This should never happen.
+		return nil, fmt.Errorf("start=%d end=%d iStart=%d iEnd=%d", start, end, iStart, iEnd)
+	}
+	return &TextMarkArray{marks: ma.marks[iStart:iEnd]}, nil
+}
+
+// BBox returns the smallest axis-aligned retangle that encloses all the TextMarks in `ma`.
```
[the remark this answers is not in the document]
Comment (peterwilliams97) [reply extractor/text.go]: Done

File: extractor/text.go [hunk 5]
```diff
@@ -848,38 +912,254 @@ func (pt PageText) length() int {
...
+type TextMark struct {
+	// Text is the extracted text. It has been decoded to Unicode via ToUnicode().
+	Text string
+	// Original is the text in the PDF. It has not been decoded like `Text`.
+	Original string
+	// BBox is the bounding box of the text.
+	BBox model.PdfRectangle
+	// Font is the font the text was drawn with.
+	Font *model.PdfFont
+	// Font is the font the text was drawn with.
```
[the remark this answers is not in the document]
Comment (peterwilliams97) [reply extractor/text.go]: Done

File: extractor/text.go [hunk 6]
```diff
@@ -897,25 +1177,36 @@ func (pt *PageText) sortPosition(tol float64) {
...
 }
 
 // toLines returns the text and positions in `pt.marks` as a slice of textLine.
 // NOTE: Caller must sort the text list top-to-bottom, left-to-right (for orientation adjusted so
 // that text is horizontal) before calling this function.
 func (pt PageText) toLines(tol float64) []textLine {
-	// We divide `pt.marks` into slices which contain texts with the same orientation, extract the lines
-	// for each orientation then return the concatention of these lines sorted by orientation.
+	// We divide `pt.marks` into slices which contain texts with the same orientation, extract the
+	// lines for each orientation then return the concatentation of these lines sorted by orientation.
```
[the remark this answers is not in the document]
Comment (peterwilliams97) [reply extractor/text.go]: Done

File: extractor/text.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (adrg) [reply extractor/text.go]: The suggested change was more about code style than performance. Repeating the same expression multiple times is usually an indication that the result could be saved in a variable. Also `lenMarks` is a bit shorter than `len(ma.marks)`.

Regarding performance, `len` is still a function so there is a bit of overhead anyway. Not a big deal here as there should not be that many characters. Also, the compiler may do optimizations on some parts of the code anyway.

File: extractor/text.go [hunk 7]
```diff
@@ -848,38 +912,255 @@ func (pt PageText) length() int {
...
+		meta = " *M*"
+	}
+	return fmt.Sprintf("{TextMark: %d %q=%02x (%5.1f, %5.1f) (%5.1f, %5.1f) %s%s}",
+		tm.Offset, tm.Text, []rune(tm.Text), b.Llx, b.Lly, b.Urx, b.Ury, font, meta)
 }
 
-// ToText returns the contents of `pt` as a single string.
-func (pt PageText) ToText() string {
+// computeViews processes the page TextMarks sorting by position and populates `pt.viewText` and
+// `pt.viewMarks` which represent the text and marks in the order which it is read on the page.
```
Comment (gunnsth) [inline extractor/text.go]: This is a breaking change (changing already exported signature), need to leave the ToText() string

Comment (gunnsth) [reply extractor/text.go]: Can leave ToText() which returns Text() and a note that ToText will be deprecated in next major version.

File: extractor/text_test.go [hunk 8]
```diff
@@ -346,3 +896,137 @@ func sortedKeys(m map[int][]string) []int {
...
+	os.Mkdir(markupDir, 0777)
+	outPath := filepath.Join(markupDir, filepath.Base(l.inPath))
+	ext := path.Ext(outPath)
+	metaPath := outPath[:len(outPath)-len(ext)] + ".json"
+
+	// Make a new PDF creator.
+	c := creator.New()
+
+	for _, pageNum := range l.pageNums() {
+
```
Comment (gunnsth) [inline extractor/text_test.go]: remove blank line at top of block

**Comment by peterwilliams97** [comment]
___
*[extractor/text.go, line 46 at r10](https://reviewable.io/reviews/unidoc/unipdf/109#-LjtxFjqDOyncGeXBGJy:-Ljzl3Sm1DADzwF8KvRx:b-896fix) ([raw file](https://github.com/unidoc/unipdf/blob/046f2d9c38797a7d5d4bc2ec26efa28eb44aefb5/extractor/text.go#L46)):*
<details><summary><i>Previously, gunnsth (Gunnsteinn Hall) wrote…</i></summary><blockquote>

return if err != nil prior to computeViews()?
</blockquote></details>

Done.
